File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        mydict = {
            ")" : "(",
            "]" : "[",
            "}" : "{"
        }
        mylist = []
        if s[0] == "}" or s[0] == "]" or s[0] == ")":
            return False
        for char in s:
            if char == "}" or char == ")" or char == "]":
                if len(mylist) == 0:
                    return False
                if mylist.pop() != mydict[char]:
                    return False
            else:
                mylist.append(char)
        if len(mylist) != 0:
            return False
        return True
            

        
# @lc code=end

File: solutions_java/stacks/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<>();
        map.put(']','[');
        map.put('}','{');
        map.put(')','(');

        Stack<Character> stack = new Stack<>();
        for (int i=0; i<s.length(); i++) {
            char w = s.charAt(i);
            if (w==']' || w=='}' || w==')') {
                if (stack.empty() || stack.pop() != map.get(w)) return false;
            } else {
              stack.push(w);
            }
        }
        if (stack.empty()) return true;
        return false;
    }
}
// @lc code=end

